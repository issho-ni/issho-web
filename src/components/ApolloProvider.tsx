import { ApolloProvider } from "@apollo/react-common"
import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory"
import ApolloClient from "apollo-client"
import { GraphQLRequest } from "apollo-link"
import { setContext } from "apollo-link-context"
import { HttpLink } from "apollo-link-http"
import * as React from "react"
import { SessionContext } from "./SessionProvider"

export interface ApolloProviderProps {
  children: React.ReactNode
}

export interface ApolloProviderState {
  client: ApolloClient<NormalizedCacheObject>
}

class IsshoApolloProvider extends React.Component<
  ApolloProviderProps,
  ApolloProviderState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>

  constructor(props: Readonly<ApolloProviderProps>) {
    super(props)

    const httpLink = new HttpLink({
      uri: GRAPHQL_ENDPOINT,
    })

    const authLink = setContext(this.setContext)

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: authLink.concat(httpLink),
    })

    this.state = { client }
  }

  public render() {
    const { client } = this.state
    client.clearStore()

    return <ApolloProvider {...{ ...this.props, client }} />
  }

  private setContext = (_: GraphQLRequest, { headers }) => ({
    headers: {
      ...headers,
      authorization: this.context.token ? `Bearer ${this.context.token}` : "",
    },
  })
}

export { IsshoApolloProvider as ApolloProvider }
