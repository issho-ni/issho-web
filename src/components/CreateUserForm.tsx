import { Field, Form, Formik } from "formik"
import * as React from "react"
import { MutationFn } from "react-apollo"
import { Link } from "react-router-dom"
import { clearState } from "../utils/state"
import { SessionContext } from "./SessionContext"

export interface CreateUserFormProps {
  createUser: MutationFn<LoginUserResult>
}

export interface CreateUserFormState {
  error?: string
}

export interface CreateUserFormValues {
  email: string
  name: string
  password: string
}

export class CreateUserForm extends React.Component<
  CreateUserFormProps,
  CreateUserFormState
> {
  public static contextType = SessionContext
  public context!: React.ContextType<typeof SessionContext>
  public state: CreateUserFormState = {}

  public render() {
    const { error } = this.state

    return (
      <Formik<CreateUserFormValues>
        initialValues={{ email: "", name: "", password: "" }}
        onSubmit={this.handleCreateUser}
      >
        {() => (
          <Form>
            {error ? <span>{error}</span> : ""}
            <Field type="text" name="name" autoComplete="name" required />
            <Field type="email" name="email" autoComplete="email" required />
            <Field
              type="password"
              name="password"
              autoComplete="new-password"
              required
            />
            <button type="submit">Create Account</button>
            <Link to="/login">Log In</Link>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateUser = (variables: CreateUserFormValues) =>
    this.props
      .createUser({ variables })
      .then(clearState(this, ["error"]))
      .then(this.context.handleLogin)
      .catch(() =>
        this.setState({
          error: "Could not create an account with that e-mail address",
        })
      )
}
