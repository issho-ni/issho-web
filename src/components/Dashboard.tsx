import * as React from "react"
import { Logout } from "./Logout"

export class Dashboard extends React.Component {
  public render() {
    return (
      <div>
        <h2>Dashboard</h2>
        <Logout />
      </div>
    )
  }
}
