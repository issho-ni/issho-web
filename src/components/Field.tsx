import { Field, FieldAttributes } from "formik"
import * as React from "react"

export interface IsshoFieldProps {
  label?: string
}

class IsshoField extends React.Component<FieldAttributes<IsshoFieldProps>> {
  public render() {
    const { label } = this.props

    return (
      <div>
        <label htmlFor={this.props.name}>{label || this.props.name}</label>
        <Field id={this.props.name} {...this.props} />
      </div>
    )
  }
}

export { IsshoField as Field }
