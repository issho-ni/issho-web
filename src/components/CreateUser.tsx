import { Mutation } from "@apollo/react-components"
import gql from "graphql-tag"
import * as React from "react"

const CreateUserForm = React.lazy(() =>
  import("./CreateUserForm").then(module => ({
    default: module.CreateUserForm,
  }))
)

export const CREATE_USER = gql`
  mutation CreateUser($email: String!, $name: String!, $password: String!) {
    createUser(input: { email: $email, name: $name, password: $password }) {
      token
      user {
        email
        name
      }
    }
  }
`

export class CreateUser extends React.Component {
  public render() {
    return (
      <Mutation<CreateUserResult> mutation={CREATE_USER}>
        {createUser => <CreateUserForm {...{ createUser }} />}
      </Mutation>
    )
  }
}
