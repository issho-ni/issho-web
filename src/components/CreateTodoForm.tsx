import { MutationFunction } from "@apollo/react-common"
import { Form, Formik, FormikActions } from "formik"
import * as React from "react"
import { Field } from "./Field"
import { GET_TODOS } from "./TodoList"

export interface CreateTodoFormProps {
  createTodo: MutationFunction<CreateTodoResult>
}

export interface CreateTodoFormValues {
  text: string
}

export class CreateTodoForm extends React.Component<CreateTodoFormProps> {
  public render() {
    return (
      <Formik<CreateTodoFormValues>
        initialValues={{ text: "" }}
        onSubmit={this.handleCreateTodo}
      >
        {() => (
          <Form>
            <Field type="text" name="text" required />
            <button type="submit">Create Todo</button>
          </Form>
        )}
      </Formik>
    )
  }

  private handleCreateTodo = (
    variables: CreateTodoFormValues,
    actions: FormikActions<CreateTodoFormValues>
  ) =>
    this.props
      .createTodo({
        optimisticResponse: {
          __typename: "Mutation",
          createTodo: {
            __typename: "Todo",
            completedAt: null,
            createdAt: new Date(),
            id: "temp",
            ...variables,
          },
        },
        update: (proxy, { data: { createTodo } }) => {
          const data = proxy.readQuery<GetTodosResult>({ query: GET_TODOS })
          data.getTodos.push(createTodo)
          proxy.writeQuery({ query: GET_TODOS, data })
        },
        variables,
      })
      .then(() => actions.resetForm({ text: "" }))
}
