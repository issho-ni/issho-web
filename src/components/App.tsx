import * as React from "react"
import { BrowserRouter } from "react-router-dom"

const ApolloProvider = React.lazy(() =>
  import("./ApolloProvider").then(module => ({
    default: module.ApolloProvider,
  }))
)
const CreateUser = React.lazy(() =>
  import("./CreateUser").then(module => ({ default: module.CreateUser }))
)
const Dashboard = React.lazy(() =>
  import("./Dashboard").then(module => ({ default: module.Dashboard }))
)
const LoginUser = React.lazy(() =>
  import("./LoginUser").then(module => ({ default: module.LoginUser }))
)
const ProtectedRoute = React.lazy(() =>
  import("./ProtectedRoute").then(module => ({
    default: module.ProtectedRoute,
  }))
)
const SessionProvider = React.lazy(() =>
  import("./SessionProvider").then(module => ({
    default: module.SessionProvider,
  }))
)

export class App extends React.Component {
  public render() {
    return (
      <React.Suspense fallback={<h1>Loadingâ€¦</h1>}>
        <SessionProvider>
          <ApolloProvider>
            <BrowserRouter>
              <ProtectedRoute
                path="/"
                authenticated
                exact
                component={Dashboard}
              />
              <ProtectedRoute path="/join" component={CreateUser} />
              <ProtectedRoute path="/login" component={LoginUser} />
            </BrowserRouter>
          </ApolloProvider>
        </SessionProvider>
      </React.Suspense>
    )
  }
}
