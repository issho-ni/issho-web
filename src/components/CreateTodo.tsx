import gql from "graphql-tag"
import * as React from "react"
import { Mutation } from "react-apollo"

const CreateTodoForm = React.lazy(() =>
  import("./CreateTodoForm").then(module => ({
    default: module.CreateTodoForm,
  }))
)

export const CREATE_TODO = gql`
  mutation CreateTodo($text: String!) {
    createTodo(input: { text: $text }) {
      id
      text
      createdAt
      completedAt
    }
  }
`

export class CreateTodo extends React.Component {
  public render() {
    return (
      <Mutation mutation={CREATE_TODO}>
        {createTodo => <CreateTodoForm {...{ createTodo }} />}
      </Mutation>
    )
  }
}
