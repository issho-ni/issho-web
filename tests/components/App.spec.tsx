import { render, wait } from "@testing-library/react"
import * as React from "react"
import { StaticRouter } from "react-router-dom"
import { mocked } from "ts-jest/utils"
import { App } from "../../src/components/App"

jest.mock("react-router-dom", () => {
  return {
    ...jest.requireActual("react-router-dom"),
    BrowserRouter: props => <div {...props} />,
  }
})

const component = (location: string) => (
  <StaticRouter {...{ location }}>
    <App />
  </StaticRouter>
)

describe("App", () => {
  beforeEach(() => jest.resetAllMocks())

  describe("/ route", () => {
    it("renders without error", async () => {
      mocked(localStorage.getItem).mockImplementation(() =>
        JSON.stringify({
          token: "not-a-valid-jwt",
        })
      )

      const { container, getByText } = render(component("/"))
      await wait(() => getByText(/dashboard/i))

      expect(container).toMatchSnapshot()
    })
  })

  describe("/join route", () => {
    it("renders without error", async () => {
      const { container, getByText } = render(component("/join"))
      await wait(() => getByText(/create account/i, { selector: "button" }))

      expect(container).toMatchSnapshot()
    })
  })

  describe("/login route", () => {
    it("renders without error", async () => {
      const { container, getByText } = render(component("/login"))
      await wait(() => getByText(/log in/i, { selector: "button" }))

      expect(container).toMatchSnapshot()
    })
  })
})
