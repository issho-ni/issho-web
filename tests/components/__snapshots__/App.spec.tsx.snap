// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App / route renders without error 1`] = `
<div>
  <div
    style="display: none;"
  >
    <div>
      <h2>
        Dashboard
      </h2>
      <a
        href="/logout"
      >
        Log out
      </a>
    </div>
  </div>
  <h1>
    Loadingâ€¦
  </h1>
</div>
`;

exports[`App /join route renders without error 1`] = `
<div>
  <div
    style=""
  >
    <form>
      
      <div>
        <label
          for="name"
        >
          name
        </label>
        <input
          autocomplete="name"
          id="name"
          name="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="email"
        >
          email
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="password"
        >
          password
        </label>
        <input
          autocomplete="new-password"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Create Account
      </button>
      <a
        href="/login"
      >
        Log In
      </a>
    </form>
  </div>
</div>
`;

exports[`App /login route renders without error 1`] = `
<div>
  <div
    style=""
  >
    <form>
      
      <div>
        <label
          for="email"
        >
          email
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="password"
        >
          password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Log In
      </button>
      <a
        href="/join"
      >
        Create Account
      </a>
    </form>
  </div>
</div>
`;
